### PR TITLE
Maryellen 0208

### PR DESCRIPTION
**Pull request by mechaffin**
# Type of change? 

- [x] Bug fix
- [ ] New feature
- [x] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description, Context, Motivation 

- When updating, look up existing data in order to get correct DateCreated.
- Ensure DateModified is updated on create and update.
- General code cleanup of RuleApiModel, ConditionApiModel, Rule, and Condition
- Remove ETag and Id from the json stored in the DB so they are not misleading

Related Issues
https://github.com/Azure/device-telemetry-java/issues/94
https://github.com/Azure/device-telemetry-java/issues/103

**Checklist:**

- [ ] All tests passed
- [ ] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: WebService/Properties/launchSettings.json [hunk 1]
```diff
@@ -12,19 +12,19 @@
       "commandName": "IISExpress",
       "launchBrowser": true,
       "launchUrl": "http://localhost:9004/v1/status",
-      "environmentVariables": {
-        "PCS_TELEMETRY_DOCUMENTDB_CONNSTRING": "AccountEndpoint={Your Document Db endpoint};AccountKey={SECRET KEY};",
-        "PCS_STORAGEADAPTER_WEBSERVICE_URL": "http://localhost:9022/v1"
-      }
+        "environmentVariables": {
+            "PCS_TELEMETRY_DOCUMENTDB_CONNSTRING": "AccountEndpoint=https://documentdb-ajz3u.documents.azure.com:443/;AccountKey=inka1YUsfF76430S753e82bRDYdxmqQIdbzSzoxqBQwA8y3surEHKt6LKKfxURMlISxkyutPLrurIUeCw44QdA==;",
```
Comment (ppathan) [inline WebService/Properties/launchSettings.json]: undo and please refresh the connection string for documentdb from portal.

File: WebService/Properties/launchSettings.json [hunk 2]
```diff
@@ -12,19 +12,19 @@
...
     "WebService": {
       "commandName": "Project",
       "launchBrowser": true,
       "launchUrl": "http://localhost:9004/v1/status",
-      "environmentVariables": {
-        "PCS_TELEMETRY_DOCUMENTDB_CONNSTRING": "AccountEndpoint={Your Document Db endpoint};AccountKey={SECRET KEY};",
-        "PCS_STORAGEADAPTER_WEBSERVICE_URL": "http://localhost:9022/v1"
-      },
+        "environmentVariables": {
+            "PCS_TELEMETRY_DOCUMENTDB_CONNSTRING": "AccountEndpoint=https://documentdb-ajz3u.documents.azure.com:443/;AccountKey=inka1YUsfF76430S753e82bRDYdxmqQIdbzSzoxqBQwA8y3surEHKt6LKKfxURMlISxkyutPLrurIUeCw44QdA==;",
```
Comment (ppathan) [inline WebService/Properties/launchSettings.json]: undo and please refresh the connection string for documentdb from portal.

File: WebService/Startup.cs [hunk 3]
```diff
@@ -40,7 +43,25 @@ public IServiceProvider ConfigureServices(IServiceCollection services)
...
+                //        options.SerializerSettings.ContractResolver = new DefaultContractResolver
+                //        {
+                //            NamingStrategy = new CamelCaseNamingStrategy()
+                //        };
+                //        options.SerializerSettings.Converters.Add(new StringEnumConverter());
+                //        //TODO: Should actually create a date formatter for our specific date format instead of setting these date based properties
+                //        options.SerializerSettings.DateFormatHandling = Newtonsoft.Json.DateFormatHandling.IsoDateFormat;
+                //        options.SerializerSettings.DateTimeZoneHandling = Newtonsoft.Json.DateTimeZoneHandling.Utc;
+                //    }
+                //);
```
Comment (ppathan) [inline WebService/Startup.cs]: If not needed can you remove.

File: WebService/v1/Controllers/RulesController.cs [hunk 4]
```diff
@@ -58,18 +58,19 @@ public RulesController(IRules ruleService)
...
         [HttpPut("{id}")]
         public async Task<RuleApiModel> PutAsync(
             [FromRoute] string id,
-            [FromBody] JToken body)
+            [FromBody] RuleApiModel rule
+            )
         {
-            Rule updatedRule = await this.ruleService.UpdateAsync(
-                new Rule(body));
+            rule.Id = id;
```
Comment (ppathan) [inline WebService/v1/Controllers/RulesController.cs]: Do we need this?

File: WebService/Startup.cs (hunk 3 above)
Comment (mechaffin) [reply WebService/Startup.cs]: Oh yes, this was from my first check in which wasn't really ready.  I'll undo it.

File: WebService/v1/Controllers/RulesController.cs (hunk 4 above)
Comment (mechaffin) [reply WebService/v1/Controllers/RulesController.cs]: Yes, and No.  This was the simplest way to ensure that the id used in the URL (the route) was actually the item being updated.  An alternative would be to actually check that they match and then throw an exception.

File: Services/Models/Rule.cs [hunk 5]
```diff
@@ -2,120 +2,27 @@
...
-                        condition["Value"].ToString()));
-                }
-                this.Conditions = conditions;
-            }
-            catch (Exception e)
-            {
-                throw new InvalidInputException("Invalid input for rule", e);
-            }
-        }
+        [JsonIgnore]
```
Comment (jillcary) [inline Services/Models/Rule.cs]: Perhaps include a comment here explaining that ETag and Id come from the StorageAdapter document and not the rule serialized string.

File: WebService/v1/Controllers/RulesController.cs (hunk 4 above)
Comment (mechaffin) [reply WebService/v1/Controllers/RulesController.cs]: I added a comment to clarify.

File: Services/Rules.cs [hunk 6]
```diff
@@ -211,10 +211,14 @@ public class Rules : IRules
 
         public async Task<Rule> CreateAsync(Rule rule)
         {
+            // Ensure dates are correct
+            rule.DateCreated = DateTimeOffset.UtcNow.ToString(DATE_FORMAT);
```
Comment (avanikp) [inline Services/Rules.cs]: Add a null check for rule before updating it.

File: Services/Rules.cs [hunk 7]
```diff
@@ -223,16 +227,21 @@ public class Rules : IRules
 
         public async Task<Rule> UpdateAsync(Rule rule)
         {
+            // Ensure dates are correct
+            // Get the existing rule so we keep the created date correct
+            var savedRule = await GetAsync(rule.Id);
```
Comment (avanikp) [inline Services/Rules.cs]: add null check for savedRule, throw an exception if null

File: Services/Rules.cs [hunk 8]
```diff
@@ -211,10 +211,14 @@ public class Rules : IRules
...
         {
+            // Ensure dates are correct
+            rule.DateCreated = DateTimeOffset.UtcNow.ToString(DATE_FORMAT);
+            rule.DateModified = rule.DateCreated;
+
             var item = JsonConvert.SerializeObject(rule);
             var result = await this.storage.CreateAsync(STORAGE_COLLECTION, item);
 
-            Rule newRule = new Rule(JToken.Parse(result.Data));
+            Rule newRule = JsonConvert.DeserializeObject<Rule>(result.Data);
```
Comment (avanikp) [inline Services/Rules.cs]: This can throw exception, add try/catch for the block.

File: WebService/v1/Controllers/RulesController.cs [hunk 9]
```diff
@@ -58,18 +58,20 @@ public RulesController(IRules ruleService)
             }
 
             // create rule from request body
-            Rule newRule = await this.ruleService.CreateAsync(new Rule(body));
+            Rule newRule = await this.ruleService.CreateAsync(rule.ToServiceModel());
```
Comment (avanikp) [inline WebService/v1/Controllers/RulesController.cs]: Add null check for rule

File: WebService/v1/Controllers/RulesController.cs [hunk 10]
```diff
@@ -58,18 +58,20 @@ public RulesController(IRules ruleService)
...
             [FromRoute] string id,
-            [FromBody] JToken body)
+            [FromBody] RuleApiModel rule
+            )
         {
-            Rule updatedRule = await this.ruleService.UpdateAsync(
-                new Rule(body));
+            //Ensure the id on the model matches the route
+            rule.Id = id;
+            Rule updatedRule = await this.ruleService.UpdateAsync(rule.ToServiceModel());
```
Comment (avanikp) [inline WebService/v1/Controllers/RulesController.cs]: same, null check

File: Services/Rules.cs (hunk 7 above)
Comment (mechaffin) [reply Services/Rules.cs]: The get actually throws a "Not Found" exception so an extra check here isn't necessary. But to be extra defensive, I'll add it.

File: WebService/v1/Controllers/RulesController.cs [hunk 11]
```diff
@@ -48,7 +48,8 @@ public RulesController(IRules ruleService)
         [HttpPost]
         public async Task<RuleApiModel> PostAsync(
             [FromQuery] string template,
-            [FromBody] JToken body)
+            [FromBody] RuleApiModel rule
+            )
```
Comment (dluc) [inline WebService/v1/Controllers/RulesController.cs]: nit: code style

File: WebService/v1/Controllers/RulesController.cs [hunk 12]
```diff
@@ -58,18 +59,29 @@ public RulesController(IRules ruleService)
...
 
             return new RuleApiModel(newRule);
         }
 
         [HttpPut("{id}")]
         public async Task<RuleApiModel> PutAsync(
             [FromRoute] string id,
-            [FromBody] JToken body)
+            [FromBody] RuleApiModel rule
+            )
```
Comment (dluc) [inline WebService/v1/Controllers/RulesController.cs]: code style

**Comment by ppathan** [comment]
@avanikp Can you please review the changes? I can merge the PR after that.

File: Services/Rules.cs [hunk 13]
```diff
@@ -223,16 +232,30 @@ public class Rules : IRules
...
+                throw new InvalidInputException("Rule not provided.");
+            }
+
+            // Ensure dates are correct
+            // Get the existing rule so we keep the created date correct
+            var savedRule = await GetAsync(rule.Id);
+            if (savedRule == null)
+            {
+                throw new ResourceNotFoundException($"Rule {rule.Id} not found.");
+            }
```
Comment (avanikp) [inline Services/Rules.cs]: nit: new line after }

File: WebService/v1/Controllers/RulesController.cs [hunk 14]
```diff
@@ -58,18 +58,28 @@ public RulesController(IRules ruleService)
             }
 
             // create rule from request body
-            Rule newRule = await this.ruleService.CreateAsync(new Rule(body));
+            if (rule == null)
+            {
+                throw new InvalidInputException("Rule not provided in request body.");
+            }
```
Comment (avanikp) [inline WebService/v1/Controllers/RulesController.cs]: nit: new line after }

File: Services/Rules.cs [hunk 15]
```diff
@@ -123,7 +123,7 @@ public class Rules : IRules
             {
                 try
```
Comment (avanikp) [inline Services/Rules.cs]: Since you added try/catch to Deserialize, is this try/catch needed anymore ?

Comment (mechaffin) [reply Services/Rules.cs]: probably not ;)

File: Services/Rules.cs (hunk 13 above)
Comment (mechaffin) [reply Services/Rules.cs]: I'm curious as to the reasoning here. The code is related to the lines that follow.  Adding extra blank lines doesn't make it more readable.

Comment (dluc) [reply Services/Rules.cs]: you can automate pretty much all these style, I think the shortcut is CTRL+ALT+L.  It's a common code style we apply across the codebase for readability (I find it more readable too)

**Pull request merged**